Authorize improvements

jordan-brough opened the pull request:
**Note:**  Please ignore the "Better usage of Payment#number" commit.  It has its own PR (#22).

It's probably easiest to understand this going through commit by commit.

Main items:
- Better error handling for authorizations
- Starting switching over to use [Amazon's Ruby gem](https://github.com/amzn/login-and-pay-with-amazon-sdk-ruby/) for performing `authorize` calls.

Secondary item:
- Allow using [Sandbox Simulations](https://payments.amazon.com/documentation/lpwa/201956480) for simulating errors while in Sandbox mode.  See code comments.

Also:
- Remove some duplicated specs
- Refactor amazon_spec.rb a bit

jordan-brough committed:
Authorization improvements: better error handling, use Amazon Ruby gem

- Handle 5xx errors
- More detailed error information
- Start using Amazon's Ruby gem while we're in the area

jordan-brough committed:
Refactor gateway/amazon_spec.rb a bit

- Consolidate gateway creation
- Use a real AmazonMws object instead of a stub

jordan-brough committed:
Remove duplicated spec code

These specs already exist (implemented slightly differently) earlier
in the file.

jordan-brough committed:
Allow using Authorization Simulation strings in Sandbox mode

By setting an appropriate ship address name on the buyer.  See code
comments.

manmartinez commented:
@jordan-brough can we rebase this please?

jordan-brough commented:
@manmartinez sure thing, I've rebased on the latest master.

jordan-brough commented:
@manmartinez just FYI I've got some more stuff in the pipeline building on this that I haven't PR'd yet. When you have a minute to take a look that would be great. Thanks!

manmartinez commented:
@jordan-brough I'll try to review this ASAP sorry for the delay but I've been caught up in another project

manmartinez commented on app/models/spree/gateway/amazon.rb:
```diff
@@ -70,22 +70,61 @@ def authorize(amount, amazon_checkout, gateway_options={})
-      response = @mws.authorize(authorization_reference_id, amount / 100.0, order.currency)
-      if response["ErrorResponse"]
-        return ActiveMerchant::Billing::Response.new(false, response["ErrorResponse"]["Error"]["Message"], response)
+      response = nil
```
Is this assignment necessary?

jordan-brough replied:
It's not. I thought it made the code more obvious/readable but I'm happy to remove it if you'd like.

jordan-brough replied:
Actually, sorry, I was reading that wrong. I think that might have been a leftover of a previous iteration of mine. I've updated to use `response = begin ...` which I think is even better and avoid the nil assignment.  Thanks.

manmartinez commented on app/models/spree/gateway/amazon.rb:
```diff
@@ -168,5 +207,40 @@ def random_suffix
+    #
+    # See https://payments.amazon.com/documentation/lpwa/201956480 for more
+    # details on Amazon Payments Sandbox Simulations.
+    def sandbox_authorize_simulation_string(order)
```
We should probably also return `nil` if `order.ship_address` is `nil`

jordan-brough replied:
Good call. Done.

manmartinez commented on app/models/spree/gateway/amazon.rb:
```diff
@@ -168,5 +207,40 @@ def random_suffix
+      minutes ||= '1'
+
+      case reason
+      when 'InvalidPaymentMethodHard' then %({"SandboxSimulation": {"State":"Declined", "ReasonCode":"InvalidPaymentMethod", "PaymentMethodUpdateTimeInMins":#{minutes}}})
```
These lines are way too long, can we try and keep them under 80 characters?

jordan-brough replied:
Hm, they are long strings that I copy-pasted from Amazon's docs. I'm with you on shooting for 80 chars but any other arrangement I tried seemed like they made the code harder to read and/or seemed excessively verbose just for the goal of 80 chars.  Any suggestions though?

manmartinez commented on lib/amazon_mws.rb:
```diff
@@ -66,15 +69,16 @@ def confirm_order
-    })
+  def authorize(ref_number, total, currency, seller_authorization_note: nil)
+    client.authorize(
+      @number,
```
~~What is `@number`? I can't remember~~ Oh it's the order number, do you think that should be a param in `authorize` and not so much an instance variable? It seems to me it makes more sense as part of the parameters of the method, also it should definitely be called `order_number` or something more verbose and not just `number` But we can revisit that in a separate PR. There's a lot of refactoring that we need in general in all the extension

jordan-brough replied:
> Oh it's the order number

Well, it's the Amazon order_reference_id, is that what you mean?  (Not the `spree_orders.number`).  I've gone ahead and renamed `@number` to `amazon_order_reference_id` and `ref_number` to `authorization_reference_id`, just like they do [in the amazon gem itself](https://github.com/amzn/login-and-pay-with-amazon-sdk-ruby/blob/5202d4fc42f0968c7835a28214f515bb566e836c/lib/pay_with_amazon/client.rb#L494-L496).

> There's a lot of refactoring that we need in general in all the extension

Very 👍 on that.  The whole concept of the AmazonMws object seems weird to me.  My thought was that we first work on getting all the plumbing updated inside `AmazonMws` to use the Amazon gem, and then once that's done just take a pass and try to get rid of the `AmazonMws` object entirely (or significantly clean it up) in a separate PR.

manmartinez commented on spec/models/spree/gateway/amazon_spec.rb:
```diff
@@ -38,16 +51,148 @@
+          'AuthorizationAmount.CurrencyCode' => order.currency
+        )
+      ).to_return(
+        return_values,
```
Should we remove the `,` at the end?

jordan-brough replied:
Personally I don't think so. I'm a huge fan of trailing commas.  I tried putting this on the same line above but it looked weird to me. I've changed this code also at this point now. Lmk if you think it's a big deal you'd like to revisit though.

manmartinez commented on spec/models/spree/gateway/amazon_spec.rb:
```diff
@@ -38,16 +51,148 @@
+    context 'with a 503 error' do
+      before do
+        # Without this the specs have a big pause while the Amazon gem retries
+        expect_any_instance_of(PayWithAmazon::Request).to receive(:get_seconds_for_try_count).at_least(:once).and_return(0)
```
I'm wondering if this should be `allow_any_instance_of` instead. An expectation will break if `pay_with_amazon` changes their implementation, using `allow_any_instance_of` won't break but it might make the tests slower...

jordan-brough replied:
Good call, done.
I also enabled `verify_partial_doubles` to help notify ourselves if Amazon's implementation changes

manmartinez commented on spec/models/spree/gateway/amazon_spec.rb:
```diff
@@ -38,16 +51,148 @@
+      end
+    end
+
+    describe 'sandbox simulation strings' do
```
I'd rather use `describe` only for public methods on the class, e.g. `describe "#authorize"` should we move this block as a `context` inside the `describe` block for `#authorize`?

```
describe '#authorize' do
  context "with sandbox simulation strings" do
    ...
  end
end
```

jordan-brough replied:
@manmartinez sure thing. Done.

manmartinez commented on spec/models/spree/gateway/amazon_spec.rb:
```diff
@@ -38,16 +51,148 @@
+      end
+      let(:expected_note) { '{"SandboxSimulation": {"State":"Declined", "ReasonCode":"InvalidPaymentMethod", "PaymentMethodUpdateTimeInMins":1}}' }
+
+      it 'forwards the note to Amazon' do
```
Can we re-write this test using a spy so we have 3 clear steps: setup, execution and validation?

``` ruby
allow(mws).to receive(:authorize).and_call_original
stub_auth_request(params)

payment_method.authorize(order.total, payment_source, {order_id: payment.send(:gateway_order_id)})

expect(mws).to have_received(:authorize).with(args)
```

jordan-brough replied:
Done.

jordan-brough committed:
Spec updates

As per PR comments

jordan-brough committed:
Refactor specs a bit

Make it easier to stub the default case.

jordan-brough committed:
Handle no ship address in sandbox simulation code

jordan-brough committed:
Update syntax

From PR comments

jordan-brough committed:
Rename some variables in AmazonMws

jordan-brough committed:
Change expect -> allow in specs

From PR comments.
And enable `verify_partial_doubles` to help notify ourselves if
Amazon's implementation changes.

jordan-brough commented:
@manmartinez thanks for the feedback! I've responded to all the comments. Lmk what you think. 

It'll be great to get this merged in so we can start swapping out AmazonMws for Amazon's gem!

manmartinez commented:
Looks good! :+1: Sorry for the delay, it's been a bit hard finding time to work on this
